openmc/deplete/microxs.py

```diff
@@ -0,0 +1,229 @@
+            Bounds for the energy group.
+
+        Returns
+        -------
+        MicroXS
+            Cross section data in [b]
+
+        """
+        groups = EnergyGroups(energy_bounds)
+
+        # Set up the reaction tallies
+        original_tallies = model.tallies
+        tallies = Tallies()
+        xs = {}
+        reactions, diluted_materials = cls._add_dilute_nuclides(chain_file,
+                                                                model,
+                                                                dilute_initial)
+        diluted_materials.export_to_xml('diluted_materials.xml')
+        model.materials = diluted_materials
+
+        for rxn in reactions:
+            if rxn == 'fission':
+                xs[rxn] = FissionXS(domain=reaction_domain, groups=groups, by_nuclide=True)
+            else:
+                xs[rxn] = ArbitraryXS(rxn, domain=reaction_domain, groups=groups, by_nuclide=True)
+            tallies += xs[rxn].tallies.values()
+
+        model.tallies = tallies
+
+        # create temporary run
+        with tempfile.TemporaryDirectory() as temp_dir:
+            statepoint_path = model.run(cwd=temp_dir)
+
+            with StatePoint(statepoint_path) as sp:
+                for rxn in xs:
+                    xs[rxn].load_from_statepoint(sp)
+
+        # Build the DataFrame
+        micro_xs = cls()
+        for rxn in xs:
+            df = xs[rxn].get_pandas_dataframe(xs_type='micro')
+            df.index = df['nuclide']
+            df.drop(['nuclide', xs[rxn].domain_type, 'group in', 'std. dev.'], axis=1, inplace=True)
+            df.rename({'mean': rxn}, axis=1, inplace=True)
+            micro_xs = concat([micro_xs, df], axis=1)
+
+        # Revert to the original tallies
+        model.tallies = original_tallies
+
+        return micro_xs
+
+    @classmethod
+    def _add_dilute_nuclides(cls, chain_file, model, dilute_initial):
+        chain = Chain.from_xml(chain_file)
+        reactions = chain.reactions
+        cross_sections = _find_cross_sections(model)
+        nuclides_with_data = cls._get_nuclides_with_data(cross_sections)
+        burnable_nucs = [nuc.name for nuc in chain.nuclides
+                               if nuc.name in nuclides_with_data]
+        diluted_materials = Materials()
+        for material in model.materials:
+            if material.depletable:
+                nuc_densities = material.get_nuclide_atom_densities()
+                dilute_density = 1.E-24 * dilute_initial
+                material.set_density('sum')
+                for nuc, density in nuc_densities.items():
+                    material.remove_nuclide(nuc)
+                    material.add_nuclide(nuc, density)
+                for burn_nuc in burnable_nucs:
+                    if burn_nuc not in nuc_densities:
+                        material.add_nuclide(burn_nuc,
+                                             dilute_density)
+            diluted_materials.append(material)
```

Comment:
This seems like it could be a potentially annoying side effect of calling `MicroXS.from_model`. Namely, the `Model` object that's passed in will be changed as a result of this block of code. Probably what we should do is 1) create new `Material` objects that match the original ones but have the dilute nuclides added, 2) create a new copy of `Model` that uses the same geometry but the new materials.

Reply:
I attempted implementing this but am now getting a weird error:

```python
    def test_from_model(model):
        ref_xs = MicroXS.from_csv('test_reference.csv')
>       test_xs = MicroXS.from_model(model, model.materials[0], CHAIN_FILE)

/home/ooblack/projects/openmc/tests/regression_tests/microxs/test.py:50: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/ooblack/projects/openmc/openmc/deplete/microxs.py:98: in from_model
    df = xs[rxn].get_pandas_dataframe(xs_type='micro')
/home/ooblack/projects/openmc/openmc/mgxs/mgxs.py:2112: in get_pandas_dataframe
    df['mean'] /= np.tile(densities, tile_factor)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/generic.py:11387: in __itruediv__
    return self._inplace_method(
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/generic.py:11352: in _inplace_method
    result = op(self, other)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/ops/common.py:70: in new_method
    return method(self, other)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/arraylike.py:124: in __truediv__
    return self._arith_method(other, operator.truediv)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/series.py:5636: in _arith_method
    return base.IndexOpsMixin._arith_method(self, other, op)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/base.py:1295: in _arith_method
    result = ops.arithmetic_op(lvalues, rvalues, op)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/ops/array_ops.py:222: in arithmetic_op
    res_values = _na_arithmetic_op(left, right, op)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/ops/array_ops.py:163: in _na_arithmetic_op
    result = func(left, right)
/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/computation/expressions.py:239: in evaluate
    return _evaluate(op, op_str, a, b)  # type: ignore[misc]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

op = <built-in function truediv>, op_str = '/'
a = array([1.98362959e-04, 1.04605679e-02, 1.92481982e-02, 3.95592500e-05,
       3.47830212e-06, 7.17484119e-09])
b = array([], dtype=float64)

    def _evaluate_standard(op, op_str, a, b):
        """
        Standard evaluation.
        """
        if _TEST_MODE:
            _store_test_result(False)
>       return op(a, b)
E       ValueError: operands could not be broadcast together with shapes (6,) (0,)

/home/ooblack/miniconda3/envs/openmc-env/lib/python3.10/site-packages/pandas/core/computation/expressions.py:69: ValueError

```

It's unclear how my change is effecting the `mgxs` side of things. I'll poke around tomorrow to see if I can fix it, but putting it here in the meantime in case anyone want's to take a crack at it.

Reply:
I think the problem may be something like the following:
- `reaction_domain` is set to `model.materials[0]`
- Under the hood, we change model.materials to the new materials with dilute nuclides
- `reaction_domain` is passed to `FissionXS`, which then tries to determine the nuclides based on that material (i.e., `model.materials[0]`, not the new material with dilute nuclides)
- Thus, the tallies that are created don't include all the nuclides we need

Reply:
subtle point! I'll keep this in mind.

Reply:
So my workaround for this is just to create a `deepcopy` of the original materials object, do all the dilution to the `Materials` object like I did before, and then once the openmc simulation is done, set `model.materials` to the deepcopy.